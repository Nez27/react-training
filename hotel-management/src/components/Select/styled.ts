import styled from 'styled-components';

const StyledSelect = styled.select`
  font-size: var(--fs-sm-x);
  padding: 5px 10px;
  border: 1px solid var(--border-color);
  border-radius: var(--radius-sm);
  font-weight: 500;
`;

export { StyledSelect };
