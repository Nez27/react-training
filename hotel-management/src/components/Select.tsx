import styled from 'styled-components';

interface ISelect {
  options: {
    value: string;
    label: string;
  }[];
  value: string;
  onChange: React.ChangeEventHandler<HTMLSelectElement>;
}

const StyledSelect = styled.select`
  font-size: var(--fs-sm-x);
  padding: 5px 10px;
  border: 1px solid var(--border-color);
  border-radius: var(--radius-sm);
  font-weight: 500;
`;

function Select({ options, value, onChange }: ISelect) {
  return (
    <StyledSelect value={value} onChange={onChange}>
      {options.map((option) => (
        <option value={option.value} key={option.value}>
          {option.label}
        </option>
      ))}
    </StyledSelect>
  );
}

export default Select;
