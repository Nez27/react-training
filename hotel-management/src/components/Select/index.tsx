import {
  FieldValues,
  RegisterOptions,
  UseFormRegister,
  useFormContext,
} from 'react-hook-form';
import { StyledSelect } from './styled';
import React, { ReactNode } from 'react';

export interface ISelectOptions {
  value: string;
  label: string;
}
interface ISelect {
  options: ISelectOptions[];
  optionsConfigForm?: RegisterOptions<FieldValues, string> | undefined;
  value?: string;
  onChange?: React.ChangeEventHandler<HTMLSelectElement> | undefined;
  id?: string;
  ariaLabel: string;
}

interface IRender extends ISelect {
  register: UseFormRegister<FieldValues>;
  children: ReactNode[];
}

const RenderSelect = ({
  options,
  id,
  value,
  optionsConfigForm,
  onChange,
  register,
  children,
  ariaLabel,
}: IRender) => {
  if (options && !register) {
    return (
      <StyledSelect
        aria-label={ariaLabel}
        id={id}
        value={value}
        onChange={onChange}
      >
        {children}
      </StyledSelect>
    );
  }

  return (
    <StyledSelect
      aria-label={ariaLabel}
      id={id}
      value={value}
      {...register(id!, optionsConfigForm)}
    >
      {children}
    </StyledSelect>
  );
};

const Select = ({
  options,
  id,
  optionsConfigForm,
  value,
  onChange,
  ariaLabel
}: ISelect) => {
  const { register } = useFormContext() ?? {};

  if (!options) return;

  return (
    <RenderSelect
      register={register}
      options={options}
      optionsConfigForm={optionsConfigForm}
      id={id}
      value={value}
      onChange={onChange}
      ariaLabel={ariaLabel}
    >
      {options.map((option) => (
        <option value={option.value} key={option.value}>
          {option.label}
        </option>
      ))}
    </RenderSelect>
  );
};

export default Select;
