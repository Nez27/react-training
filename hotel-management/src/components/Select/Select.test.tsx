import { render } from '@testing-library/react';

// Components
import Select from '.';

describe('Select', () => {
  const options = [
    {
      label: 'Option 1',
      value: 'Value 1',
    },
    {
      label: 'Option 2',
      value: 'Value 2',
    },
  ];
  const value = 'Temp value';
  const handleOnChange = jest.fn();

  const wrapper = render(
    <Select
      options={options}
      value={value}
      onChange={handleOnChange}
      ariaLabel="Sort"
    />
  );

  test('Should render correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
