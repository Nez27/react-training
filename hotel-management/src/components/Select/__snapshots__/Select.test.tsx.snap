// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select Should render correctly 1`] = `
<select
  aria-label="Sort"
  className="sc-bdnyFh dqXTRP"
  onChange={[MockFunction]}
  value="Temp value"
>
  <option
    value="Value 1"
  >
    Option 1
  </option>
  <option
    value="Value 2"
  >
    Option 2
  </option>
</select>
`;
